File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, cx, cy, itemX, itemY):
    answer = 0
    maps = [[-1]*102 for _ in range(102)]
    
    for rec in rectangle:
        lx, ly, rx, ry = map(lambda x:x*2, rec)
        for i in range(lx, rx+1):
            for j in range(ly, ry+1):
                if lx<i<rx and ly<j<ry:
                    maps[i][j] = 0
                elif maps[i][j] == -1:
                    maps[i][j] = 1
                    
    q = deque([[cx*2, cy*2]])
    visit = [[0]*102 for _ in range(102)]
    visit[cx*2][cy*2] = 1
    dir = [[0,1], [0,-1], [1,0], [-1,0]]
    while q:
        x, y = q.popleft()
        if x == itemX*2 and y == itemY*2:
            answer = visit[x][y]//2
            break
            
        for di, dj in dir:
            ni, nj = x+di, y+dj
            if visit[ni][nj] == 0 and maps[ni][nj] == 1:
                q.append([ni,nj])
                visit[ni][nj] = visit[x][y] + 1
    
    return answer

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.js
function solution(rectangle, characterX, characterY, itemX, itemY) {
    var answer = 0;
    let board = Array.from(Array(102), ()=>Array(102).fill(-1));
    rectangle.forEach((el, i)=> {
        el.forEach((e, j)=> rectangle[i][j] = e*2);
    });
    
    rectangle.forEach((el) => {
        let x1 = el[0], y1 = el[1], x2 = el[2], y2 = el[3];
        for(let x=x1; x<=x2; x++){
            for(let y=y1; y<=y2; y++){
                if (x1<x&&x<x2 && y1<y&&y<y2) board[x][y] = 0;
                else if (board[x][y] === -1) board[x][y] = 1;
            }
        }
    });
    
    let q = [[characterX*2, characterY*2]];
    let visit = Array.from(Array(102), ()=>Array(102).fill(0));
    visit[characterX*2][characterY*2] = 1;
    let dir = [[0,1], [0,-1], [1,0], [-1,0]];
    while (q.length > 0) {
        let now = q.shift();
        let x = now[0], y = now[1];
        if (x === itemX*2 && y === itemY*2){
            answer = (visit[x][y] - 1) / 2;
            break;
        }
        for (d of dir){
            let nx = x+d[0], ny = y+d[1];
            if (visit[nx][ny] === 0 && board[nx][ny] === 1) {
                q.push([nx, ny]);
                visit[nx][ny] = visit[x][y] + 1;
            }
        }
    }
    
    
    return answer;
}